### PR TITLE
Add dev shell for the controller

### PR DESCRIPTION
**Pull request by steshaw**
The default dev shell still has the base packages (such as poetry, terraform, kubectl, etc) — those that were originally in the `shell.nix`. The controller dev shell has all the base packages along with the `mkPoetryEnv` for controller as well as controller itself (to bring in the `run_rexfw_mpi` script).

The kludge of setting `LD_LIBRARY_PATH` in the shell is no longer required.

## Developer workflow with the new dev shell

The workflow for working with the controller in local development mode goes something like this:

1. Enter the dev shell with `nix develop .#controller`
2. Edit `./lib/runners/rexfw/pyproject.toml` to add dependencies necessary to run the `probability.py`. e.g.

```toml
chainsail-helpers = "*"
scipy = "*"
pymc = "*"
```

3. Update the `poetry.lock` for the controller to bring in the additional dependencies:

```sh-session
$ cd ./app/controller
$ poetry lock --no-update
```

4. Exit and re-enter the controller dev shell
5. You should find that you have a Python interpreter on your PATH with the necessary dependencies (note the `-env`):

```sh-session
$ which python
/nix/store/4834b5pqk6fsn1sjh1mrpwdln9jw5nrj-python3-3.10.9-env/bin/python
```

6. Then, assuming you have [chainsail-resources](https://github.com/tweag/chainsail-resources/) cloned at the same level as [chainsail](https://github.com/tweag/chainsail) and you have a job.json in the `./app/controller` directory such as  [job.json](https://gist.github.com/steshaw/10edb377b89a5c315d92c3c2e40454ea), you can run the examples from `chainsail-resources` except for httpstan (httpstan needs an extra server running which I haven't been able to connect with the chainsail-controller-local as yet).

```sh-session
$ cd ./app./controller
$ PYTHONPATH=../../../chainsail-resources/examples/mixture/ chainsail-controller-local --job-spec=job.json --dirname=/tmp/out
```

```sh-session
$ PYTHONPATH=../../../chainsail-resources/examples/pymc-mixture/ chainsail-controller-local --job-spec=job.json --dirname=/tmp/out
```

```sh-session
$ PYTHONPATH=../../../chainsail-resources/examples/soft-kmeans/ chainsail-controller-local --job-spec=job.json --dirname=/tmp/out
```

7. You can edit local source files and they will be reflected the next time you run `chainsail-controller-local`.

This workflow is now documented [as part of the PR](https://github.com/tweag/chainsail/blob/31efc13759be5a11ba089ff96579fe7899943fc0/app/controller/README.md#development-workflow).

### Direnv

I added direnv support which means that you can use `direnv allow` to get a controller dev shell. This also allows VS Code Python tooling to work along with the [direnv-vscode](https://marketplace.visualstudio.com/items?itemName=mkhl.direnv). The VSCode Python extension is Pylance from Microsoft which is powered by Pyright.

### CODE REVIEW

File: app/controller/pyproject.toml
Comment (steshaw) [inline app/controller/pyproject.toml:13]: This subtle change is necessary to make `controller` compatible with `chainsail-helpers` (as it has a similar <3.11 constraint).

File: flake.nix
Comment (steshaw) [inline flake.nix:38]: I discovered that it does seem to work to set `preferWheels` here. I assume that poetry2nix does some kind of post-processing of the precompiled `.whl` files that it downloads.

**Review by simeoncarstens — APPROVED** [review]
Looks good, thanks! Happy to merge as is, but I made a couple suggestions anyways :slightly_smiling_face:

File: flake.nix
Comment (steshaw) [reply flake.nix:38]: It defaults to `false`. My understanding is that it downloads sources (and sdist in Python parlance) and builds them.

File: .envrc
Comment (steshaw) [reply .envrc:1]: I usually find that folks do commit their `.envrc` files. If folks are not using `direnv`, it will not affect them. If they do use `direnv`, but do not have Nix installed, nothing happens until they do `direnv allow` which would then try to enter the shell and fail. Then can then do `direnv deny` to disable direnv again. So overall, it's pretty harmless. Also, if folks are using VSCode, then it could help them get a working setup.

File: flake.nix
Comment (steshaw) [reply flake.nix:33]: Yeah, it takes a bit of getting used to! It's how it looks in Jupyenv and `nixpkgs-fmt` likes it.

File: flake.nix
Comment (steshaw) [reply flake.nix:38]: FWIW, I imagine that it's safer to rebuild the package from scratch in a Nix environment. On the other hand, I recall you saying that you were surprised that something compiled when you used the dev shell some time back, so I thought I'd try this option. One of the reasons will be that when using traditional Python tooling (e.g. `pip`), it automatically favours installing wheels (precompiled dependencies). Using `preferWheels` with poetry2nix seems popular when I review other projects on GitHub (though I once noticed that someone disabled it specifically for macOS/darwin). I imagine that poetry2nix defaults to recompiling because of a safety factor. However, I don't know all the trade-offs. For instance, I don't know what kind of post-processing is done on the wheels to make them work with Nix, but I assume that some kind of post-processing is required to adjust the precompiled libraries to use Nix store paths.

We can switch it back to `preferWheels = false`. Whichever is fine with me until I am better informed on it 🤷. If we set up a Cachix cache, then devs can share the Nix-compiled dependencies (once pushed by CI), and so would obviate somewhat the desire to use precompiled wheels.

File: flake.nix
Comment (steshaw) [reply flake.nix:70]: It seems a little whimsical to change the definition to `default = pythonDevShell`. The way I see it is that the default environment is different to the ones for working on a specific service/application. i.e. it's not completely symmetrical because there's a distinction. But this is also a bit whimsical on my behalf.

What you suggest doesn't quite work in any case. We'd need to change the argument to `pythonDevShell` to be an attrset and choose a name for the argument in order to give it a default argument. Then the default shell would look like `default = pythonDevShell {}`. However, in the case of getting a working Python environment for doing work on a specific application, an argument is always required (the `mkPoetryEnv` package).

Comment (steshaw) [reply flake.nix:70]: I agree with your point that there are quite a few base packages and these could be split up in a later PR. The `pythonDevShell` is a bit of a misnomer, it could be called `completePythonDevShellWithOtherPackagesYouWillNeed` 🙂. The only downside of splitting things up to have less initial downloading is that developers may end up finding that they are in the wrong shell which would be frustrating. We should muse on it a little more and raise an issue if it needs addressing.

Some initial musings:
- A halfway measure may be to pare back some of the dependencies. For instance, I'm not sure what `file` is for.
- If we decide to Nixify the `client` front end, we might find that we can drop `yarn` from the global base packages.

**Comment by steshaw** [comment]
I forgot to call out explicitly for feedback on the developer workflow that I outlined above. @simeoncarstens, since I'm not a Python developer, I was curious about what you think of this developer workflow. Does it seem strange or unorthodox? I wondered if Python devs these days are used to running everything via `poetry run`, etc.

**Comment by steshaw** [comment]
Note that I've rebased on changes that occurred in https://github.com/tweag/chainsail/pull/446.

**Comment by steshaw** [comment]
Warning, we shouldn't merge this PR until the stack of PRs it builds on is merged first.

File: flake.nix
Comment (simeoncarstens) [reply flake.nix:41]: Thanks for these detailed explanations! If there's precedence for that option elsewhere in projects, and it works, and it might save some build time, then let's definitely leave it activated :slightly_smiling_face:

File: flake.nix
Comment (simeoncarstens) [reply flake.nix:73]: Regarding your first message: to be honest, I don't quite follow. https://github.com/tweag/chainsail/blob/controller-dev-shell/flake.nix#L63 tells me that `ps` is a list (because the `++` operator doesn't work for attribute sets. And `pkgs.mkShell { packages = basePackages; };` (which is the current `default`) is exactly the same thing as `pythonDevShell []`. That's why I suggested to replace the current `default` and to rename `pythonDevShell`.

As for splitting up the dependencies, your points are very good. But let's think about those later :slightly_smiling_face:

File: flake.nix
Comment (simeoncarstens) [inline flake.nix:65]: I now also tested your proposed developer workflow and that works smoothly! Yes, it's a couple of steps, but I don't think there is a way around this when working fully with Nix. It would be cool, though, if we could replace the `python -m chainsail.controller.run_local` incantation with something like an alias `run-controller-local`, similar to the Poetry script when you use the Poetry shell. Maybe there is an option to pass through a `shellHook` in `mkPoetryEnv`, but a quick search was inconclusive. But that shouldn't prevent merging, just a nice to have :slightly_smiling_face:

File: flake.nix
Comment (steshaw) [reply flake.nix:73]: Yes, `ps` is indeed a list. Your original comment would have made more sense to me if you had suggested `default = pythonDevShell []` and at the same time suggest a rename. My perspective was that I wanted to treat the creation of development shells for Python applications as distinct from the one that creates a generic shell with useful utilities. However, I don't think it's important either way.

File: flake.nix
Comment (steshaw) [reply flake.nix:65]: It would be straightforward to add a `shellHook` if you're interested. Not to `mkPoetryEnv`, but just in the usual place as an argument to `mkShell`.

Comment (simeoncarstens) [reply flake.nix:65]: True, in `mkShell` it's easy! Let's do that then; it would then somewhat maintain the interface of the previous way of running the controller locally.

Comment (simeoncarstens) [reply flake.nix:65]: @steshaw If you'd like to add the above `mkShell` argument, then we can merge this, I think. But even without it should be fine, but in any case we should document the Nix-based development workflow.

Comment (steshaw) [reply flake.nix:65]: Thanks for the bump here, I'd forgotten about adding an alias or script to the development environment. I'd really like to get this merged. Unfortunately, this PR is stacked on top of https://github.com/tweag/chainsail/pull/446 which is blocked by https://github.com/tweag/chainsail/issues/449. However, I was able to test an end-to-end workflow for #446 (which I assume exercises Celery tasks). To start the frontend dev server for the E2E workflow, I temporarily downgraded the nixpkgs pin to `nixpkgs/21.11`. I was able to create, start, and finish a job using the "mixture.zip". I'll make a similar note against #446 in case you are happy to merge that given that.

Comment (steshaw) [reply flake.nix:65]: It turns out that there is no need for an alias or script as the environment already includes the scripts from `pyproject.toml`!

```sh-session
$ which chainsail-controller-local
/nix/store/ax7z9q980m2mxg9s468icylb06285nwr-python3-3.10.9-env/bin/chainsail-controller-local
```

Comment (simeoncarstens) [reply flake.nix:65]: Oh nice, I didn't realize these get carried over when using `mkPoetryEnv`!

Comment (simeoncarstens) [reply flake.nix:65]: I guess that script is also correctly wrapped and everything by poetry2nix?

Comment (steshaw) [reply flake.nix:65]: Yes, exactly

**Comment by steshaw** [comment]
Unfortunately, a nasty Git conflict has arisen. I'll have time to look at it in a few days.

**Comment by steshaw** [comment]
@simeoncarstens the nasty Git conflict has been resolved — ready to merge!

**Comment by simeoncarstens** [comment]
Perfect, thanks so much for this :tada: :blush:

**Pull request merged**
